drivers/modem/wncm14a2a.c

```diff
@@ -0,0 +1,1830 @@
+
+		printk("%02X ", byte);
+
+		if (byte < 0x20 || byte > 0x7f) {
+			output[k++] = '.';
+		} else {
+			output[k++] = byte;
+		}
+
+		n++;
+		if (n % 8 == 0) {
+			if (n % 16 == 0) {
+				output[k] = '\0';
+				printk(" [%s]\n", output);
+				k = 0;
+			} else {
+				printk(" ");
+			}
+		}
+	}
+
+	if (n % 16) {
+		int i;
+
+		output[k] = '\0';
+
+		for (i = 0; i < (16 - (n % 16)); i++) {
+			printk("   ");
+		}
+
+		if ((n % 16) < 8) {
+			printk(" "); /* one extra delimiter after 8 chars */
+		}
+
+		printk(" [%s]\n", output);
+	}
+}
+#else
+#define _hexdump(...)
+#endif
+
+static struct wncm14a2a_socket *socket_get(void)
+{
+	int i;
+	struct wncm14a2a_socket *sock = NULL;
+
+	for (i = 0; i < MDM_MAX_SOCKETS; i++) {
+		if (!ictx.sockets[i].context) {
+			sock = &ictx.sockets[i];
+			break;
+		}
+	}
+
+	return sock;
+}
+
+static struct wncm14a2a_socket *socket_from_id(int socket_id)
+{
+	int i;
+	struct wncm14a2a_socket *sock = NULL;
+
+	if (socket_id < 1) {
+		return NULL;
+	}
+
+	for (i = 0; i < MDM_MAX_SOCKETS; i++) {
+		if (ictx.sockets[i].socket_id == socket_id) {
+			sock = &ictx.sockets[i];
+			break;
+		}
+	}
+
+	return sock;
+}
+
+static void socket_put(struct wncm14a2a_socket *sock)
+{
+	if (!sock) {
+		return;
+	}
+
+	sock->context = NULL;
+	sock->socket_id = 0;
+	memset(&sock->src, 0, sizeof(struct sockaddr));
+	memset(&sock->dst, 0, sizeof(struct sockaddr));
+}
+
+char *wncm14a2a_sprint_ip_addr(const struct sockaddr *addr)
+{
+	static char buf[NET_IPV6_ADDR_LEN];
+
+#if defined(CONFIG_NET_IPV6)
+	if (addr->sa_family == AF_INET6) {
+		return net_addr_ntop(AF_INET6, &net_sin6(addr)->sin6_addr,
+				     buf, sizeof(buf));
+	}
+#endif
+#if defined(CONFIG_NET_IPV4)
+	if (addr->sa_family == AF_INET) {
+		return net_addr_ntop(AF_INET, &net_sin(addr)->sin_addr,
+				     buf, sizeof(buf));
+	}
+#endif
+
+	SYS_LOG_ERR("Unknown IP address family:%d", addr->sa_family);
+	return NULL;
+}
+
+/* Send an AT command with a series of response handlers */
+static int send_at_cmd(struct wncm14a2a_socket *sock,
+			const u8_t *data, int timeout)
+{
+	int ret;
+
+	ictx.last_error = 0;
+
+	SYS_LOG_DBG("OUT: [%s]", data);
+	mdm_receiver_send(&ictx.mdm_ctx, data, strlen(data));
+	mdm_receiver_send(&ictx.mdm_ctx, "\r\n", 2);
+
+	if (timeout == K_NO_WAIT) {
+		return 0;
+	}
+
+	if (!sock) {
+		k_sem_reset(&ictx.response_sem);
+		ret = k_sem_take(&ictx.response_sem, timeout);
+	} else {
+		k_sem_reset(&sock->sock_send_sem);
+		ret = k_sem_take(&sock->sock_send_sem, timeout);
+	}
+
+	if (ret == 0) {
+		ret = ictx.last_error;
+	} else if (ret == -EAGAIN) {
+		ret = -ETIMEDOUT;
+	}
+
+	return ret;
+}
+
+static int send_data(struct wncm14a2a_socket *sock, struct net_pkt *pkt)
+{
+	int ret;
+	struct net_buf *frag;
+	char buf[sizeof("AT@SOCKWRITE=#,####,1\r")];
+
+	if (!sock) {
+		return -EINVAL;
+	}
+
+	ictx.last_error = 0;
+
+	frag = pkt->frags;
+	/* use SOCKWRITE with binary mode formatting */
+	snprintk(buf, sizeof(buf), "AT@SOCKWRITE=%d,%u,1%c",
```

Comment:
Minor thing, but you could write the \r directly to the format string and avoid %c

Reply:
Hah, yes, I'll add it directly.